src/compiler/resolutionCache.ts

```diff
@@ -0,0 +1,316 @@
+            }
+        }
+
+        function closeFailedLookupLocationWatcher(failedLookupLocation: string, failedLookupLocationPath: Path, containingFile: string, name: string) {
+            const failedLookupLocationWatcher = failedLookupLocationsWatches.get(failedLookupLocationPath);
+            Debug.assert(!!failedLookupLocationWatcher);
+            failedLookupLocationWatcher.refCount--;
+            if (failedLookupLocationWatcher.refCount) {
+                log(`Watcher: FailedLookupLocations: Status: Removing existing watcher: Location: ${failedLookupLocation}, containingFile: ${containingFile}, name: ${name}: refCount: ${failedLookupLocationWatcher.refCount}`);
+            }
+            else {
+                log(`Watcher: FailedLookupLocations: Status: Closing the file watcher: Location: ${failedLookupLocation}, containingFile: ${containingFile}, name: ${name}`);
+                failedLookupLocationWatcher.fileWatcher.close();
+                failedLookupLocationsWatches.delete(failedLookupLocationPath);
+            }
+        }
+
+        type FailedLookupLocationAction = (failedLookupLocation: string, failedLookupLocationPath: Path, containingFile: string, name: string) => void;
+        function withFailedLookupLocations(failedLookupLocations: string[], containingFile: string, name: string, fn: FailedLookupLocationAction) {
+            forEach(failedLookupLocations, failedLookupLocation => {
+                fn(failedLookupLocation, toPath(failedLookupLocation), containingFile, name);
+            });
+        }
+
+        function updateFailedLookupLocationWatches(containingFile: string, name: string, existingFailedLookupLocations: string[], failedLookupLocations: string[]) {
+            if (failedLookupLocations) {
+                if (existingFailedLookupLocations) {
+                    const existingWatches = arrayToMap(existingFailedLookupLocations, failedLookupLocation => toPath(failedLookupLocation));
+                    for (const failedLookupLocation of failedLookupLocations) {
+                        const failedLookupLocationPath = toPath(failedLookupLocation);
+                        if (existingWatches && existingWatches.has(failedLookupLocationPath)) {
+                            // still has same failed lookup location, keep the was
+                            existingWatches.delete(failedLookupLocationPath);
+                        }
+                        else {
+                            // Create new watch
+                            watchFailedLookupLocation(failedLookupLocation, failedLookupLocationPath, containingFile, name);
+                        }
+                    }
+
+                    // Close all the watches that are still present in the existingWatches since those are not the locations looked up for buy new resolution
+                    existingWatches.forEach((failedLookupLocation, failedLookupLocationPath: Path) =>

```

Comment:
Annotating a parameter type acts like a cast due to function bivariance. I would prefer a regular `as Path` cast.